services/worker-manager/src/provisioner.js

```diff
@@ -96,26 +104,79 @@ class Provisioner {
    * Run a single provisioning iteration
    */
   async provision() {
+    if (this.provisioningLoopAlive) {
+      this.monitor.notice('loop-interference', {});
+      return;
+    }
+    this.provisioningLoopAlive = true;
     // Any once-per-loop work a provider may want to do
     await this.providers.forAll(p => p.prepare());
 
+    // track the providerIds seen for each worker pool, so they can be removed
+    // from the list of previous provider IDs
+    const providersByPool = new Map();
+    const seen = (providerId, workerPoolId, workerCapacity) => {
+      const v = providersByPool.get(workerPoolId);
+      if (v) {
+        v.providers.add(providerId);
+        v.count += workerCapacity;
+      } else {
+        providersByPool.set(workerPoolId, {providers: new Set([providerId]), count: workerCapacity});
+      }
+    };
+
+    // Check the workers
+    await this.providers.forAll(p => p.scanPrepare());
+    await this.Worker.scan({
+      state: Entity.op.notEqual(this.Worker.states.STOPPED),
+    }, {
+      handler: async worker => {
+        seen(worker.providerId, worker.workerPoolId, worker.capacity);
```

Comment:
Oh, this is really neat!  If we change a worker pool's provider, the new provider will only provision capacity not already supported by workers from the old provider.  So it'd be possible to, for example, transition `gecko-t/t-linux-xlarge` from one AWS account to another with no downtime and no over-provisioning.  How cool is that?